### PR TITLE
feat: support for tools in `HuggingFaceAPIChatGenerator`

### PR DESCRIPTION
**Pull request by anakin87**
### Related Issues

- fixes https://github.com/deepset-ai/haystack-private/issues/85

### Proposed Changes:
I'm adding support for tools in `HuggingFaceAPIChatGenerator`. Similar to OpenAI PR https://github.com/deepset-ai/haystack-experimental/pull/57.

### How did you test it?
CI, new tests.

### Notes for the reviewer
- Don't be frightened by this large PR. Most of the code is tests.

- Support for tools in the HF API/TGI is still experimental and not fully polished. This means the model in its own could work better than called via HF API/run in a TGI container. Examples: the model, if provided with tools, always tend to prepare tool calls; some models (Mistral) are not completely well supported. 
  - For more details you can read: https://github.com/deepset-ai/haystack-private/issues/85 https://github.com/huggingface/huggingface_hub/issues/2606 https://github.com/huggingface/text-generation-inference/issues/2461 https://github.com/huggingface/text-generation-inference/issues/2549

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

### CODE REVIEW

File: haystack_experimental/components/generators/chat/hugging_face_api.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+        if tools:
+            tool_names = [tool.name for tool in tools]
+            duplicate_tool_names = {name for name in tool_names if tool_names.count(name) > 1}
+            if duplicate_tool_names:
+                raise ValueError(f"Duplicate tool names found: {duplicate_tool_names}")
+        self.tools = tools
+
+        if tools and streaming_callback is not None:
+            raise ValueError("Using tools and streaming at the same time is not supported. Please choose one.")
```
Comment (anakin87) [inline haystack_experimental/components/generators/chat/hugging_face_api.py:190]: The HF API allows this use case, but I made this decision for the following reasons:
- the streaming output to expect when there are tool calls is pratically undocumented
- based on my experiments, I see that depending on the model, the API returns differend end tokens (e.g. `</s>`, <|eot_id|>...).
- even if some models (Llama 3.1) support producing multiple tool calls in the same interaction, I could not reproduce this with HF API and I could not infer the format to expect.

In short, on the HF API side, it seems to me that there is work to be done in terms of standardization and documentation. We may support tools+streaming in the future, when things are clearer...

File: haystack_experimental/components/generators/chat/hugging_face_api.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+        # the base class __init__ also checks the hugingface_hub lazy import
+        super(HuggingFaceAPIChatGenerator, self).__init__(
+            api_type=api_type,
+            api_params=api_params,
+            token=token,
+            generation_kwargs=generation_kwargs,
+            stop_words=stop_words,
+            streaming_callback=streaming_callback,
+        )
```
Comment (shadeMe) [inline haystack_experimental/components/generators/chat/hugging_face_api.py:200]: This should be the first statement in the init method.

File: haystack_experimental/components/generators/chat/hugging_face_api.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+        if tools and self.streaming_callback is not None:
+            raise ValueError("Using tools and streaming at the same time is not supported. Please choose one.")
+
+        hf_tools = None
+        if tools:
+            hf_tools = [{"type": "function", "function": {**t.tool_spec}} for t in tools]
+
+        if self.streaming_callback:
+            return self._run_streaming(formatted_messages, generation_kwargs, hf_tools)
```
Comment (shadeMe) [inline haystack_experimental/components/generators/chat/hugging_face_api.py:268]: Why do we need to pass the tools if they don't support streaming?

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: test/components/generators/chat/test_hugging_face_api.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+                "name": "name",
+                "parameters": {
+                    "x": {
+                        "type": "string",
+                    },
+                },
+            }
+        ]
+
+    def test_from_dict(self, mock_check_valid_model):
```
[the remark this answers is not in the document]
Comment (anakin87) [reply test/components/generators/chat/test_hugging_face_api.py:287]: added

File: haystack_experimental/components/generators/chat/hugging_face_api.py (hunk 2 above)
Comment (anakin87) [reply haystack_experimental/components/generators/chat/hugging_face_api.py:190]: fixed

File: haystack_experimental/components/generators/chat/hugging_face_api.py (hunk 3 above)
Comment (anakin87) [reply haystack_experimental/components/generators/chat/hugging_face_api.py]: Right! This is a leftover of my experiments.
Removed.

File: test/components/generators/chat/test_hugging_face_api.py [hunk 5]
```diff
@@ -305,6 +306,30 @@ def test_from_dict(self, mock_check_valid_model):
...
+        pipeline.add_component("generator", generator)
+
+        pipeline_dict = pipeline.to_dict()
+        print(pipeline_dict)
+
+        pipeline_yaml = pipeline.dumps()
+        print(pipeline_yaml)
+
+        # new_pipeline = Pipeline.loads(pipeline_yaml)
+        # assert new_pipeline==pipeline
```
Comment (shadeMe) [inline test/components/generators/chat/test_hugging_face_api.py:331]: This should be uncommented.

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: test/components/generators/chat/test_hugging_face_api.py (hunk 5 above)
Comment (anakin87) [reply test/components/generators/chat/test_hugging_face_api.py]: yes, sorry!
